lib/hawkbit/hawkbit.c

```diff
@@ -0,0 +1,1273 @@
+			    JSON_TOK_STRING),
+	JSON_OBJ_DESCR_PRIM(struct hawkbit_dep_res_chunk, name,
+			    JSON_TOK_STRING),
+	JSON_OBJ_DESCR_OBJ_ARRAY(struct hawkbit_dep_res_chunk, artifacts,
+				HAWKBIT_DEP_MAX_CHUNK_ARTS, num_artifacts,
+				json_dep_res_arts_descr,
+				ARRAY_SIZE(json_dep_res_arts_descr)),
+};
+
+static const struct json_obj_descr json_dep_res_deploy_descr[] = {
+	JSON_OBJ_DESCR_PRIM(struct hawkbit_dep_res_deploy, download,
+			    JSON_TOK_STRING),
+	JSON_OBJ_DESCR_PRIM(struct hawkbit_dep_res_deploy, update,
+			    JSON_TOK_STRING),
+	JSON_OBJ_DESCR_OBJ_ARRAY(struct hawkbit_dep_res_deploy, chunks,
+				 HAWKBIT_DEP_MAX_CHUNKS, num_chunks,
+				 json_dep_res_chunk_descr,
+				 ARRAY_SIZE(json_dep_res_chunk_descr)),
+};
+
+static const struct json_obj_descr json_dep_res_descr[] = {
+	JSON_OBJ_DESCR_PRIM(struct hawkbit_dep_res, id, JSON_TOK_STRING),
+	JSON_OBJ_DESCR_OBJECT(struct hawkbit_dep_res, deployment,
+			      json_dep_res_deploy_descr),
+};
+
+static const struct json_obj_descr json_dep_fbk_descr[] = {
+	JSON_OBJ_DESCR_PRIM(struct hawkbit_dep_fbk, id, JSON_TOK_STRING),
+	JSON_OBJ_DESCR_OBJECT(struct hawkbit_dep_fbk, status,
+			      json_status_descr),
+};
+
+static bool start_http_client()
+{
+	struct addrinfo *addr;
+	struct addrinfo hints;
+	int resolve_attempts = 10;
+	int ret = -1;
+
+	if (IS_ENABLED(CONFIG_NET_IPV6)) {
+		hints.ai_family = AF_INET6;
+		hints.ai_socktype = SOCK_STREAM;
+	} else if (IS_ENABLED(CONFIG_NET_IPV4)) {
+		hints.ai_family = AF_INET;
+		hints.ai_socktype = SOCK_STREAM;
+	}
+
+#if defined(CONFIG_NET_SOCKETS_SOCKOPT_TLS)
+	int protocol = IPPROTO_TLS_1_2;
+	char port[] = CONFIG_HAWKBIT_PORT;
+#else
+	int protocol = IPPROTO_TCP;
+	char port[] = CONFIG_HAWKBIT_PORT;
+#endif
+	while (resolve_attempts--) {
+		ret = getaddrinfo(HAWKBIT_SERVER, port, &hints, &addr);
+		if (ret == 0) {
+			break;
+		}
+		k_sleep(K_SECONDS(1));
+	}
+	if (ret != 0) {
+		LOG_ERR("Could not resolve dns");
+		return false;
+	}
+
+	hb_context.sock = socket(addr->ai_family, SOCK_STREAM, protocol);
+	if (hb_context.sock < 0) {
+		LOG_ERR("Failed to create TCP socket");
+		return false;
+	}
+
+#if defined (CONFIG_NET_SOCKETS_SOCKOPT_TLS)
+        sec_tag_t sec_tag_opt[] = {
+                CA_CERTIFICATE_TAG,
+        };
+        if (setsockopt(hb_context.sock, SOL_TLS, TLS_SEC_TAG_LIST,
+                         sec_tag_opt, sizeof(sec_tag_opt)) < 0) {
+		LOG_ERR("Failed to set TLS_TAG option");
+		return false;
+	}
+
+        if (setsockopt(hb_context.sock, SOL_TLS, TLS_HOSTNAME,
+                         HAWKBIT_SERVER, sizeof(HAWKBIT_SERVER)) < 0) {
+		return false;
+	}
+#endif
+
+	if (connect(hb_context.sock, addr->ai_addr, addr->ai_addrlen) < 0) {
+		LOG_ERR("Failed to connect to Server");
+		return false;
+	}
+
+	return true;
+}
+
+static void cleanup_connection(void)
+{
+	if (close(hb_context.sock) < 0) {
+		LOG_ERR("Could not close the socket");
+	}
+
+	hb_context.sock = 0;
+}
+
+static int atoi_n(const char *s, int len)
```

Comment:
`atoi`, in general, is deprecated. Use `strtol` instead (if you feel that bounds checking is absolutely necessary here, add `strntol`).

Reply:
> atoi, in general, is deprecated. Use strtol instead

And my 2 cents: deprecated or not, but duplicating adhoc implementation of such functions throughout subsystems definitely not ok. So, there's 2 choices: a) use already existing functions (as e.g. available in minlibc, if newlib is considered to heavy); b) contribute missing functions to minlibc (assuming it will be accepted; I personally ready to give +1 for implementation of any standard function, as long as it's useful for porting some software).